display_list/display_list_builder.cc

```diff
@@ -709,6 +709,21 @@ SkRect DisplayListBuilder::getLocalClipBounds() {
   return kMaxCullRect_;
 }
 
+bool DisplayListBuilder::quickReject(const SkRect& bounds) const {
+  if (bounds.isEmpty()) {
+    return true;
+  }
+  SkMatrix matrix = getTransform();
+  // We don't need the inverse, but this method tells us if the matrix
+  // is singular in which case we can reject all rendering.
+  if (!matrix.invert(nullptr)) {
```

Comment:
I should have said "for a leaf 2D operation". All you lose with the 3x3 is composing with future 4x4 matrices and ability to take or generate Z components. None of our geometry has Z and we don't output Z so the only thing that matters is the composability.

That composability is critical for TransformLayer because it could be nested with another TransformLayer, but not in single layer operations.